crates/cli/src/main.rs

```diff
@@ -27,72 +31,90 @@ use futures::{
-        io
-    };
-
-    std::thread::spawn(move || {
-        let runtime = tokio::runtime::Builder::new_multi_thread()
-            .enable_all()
-            .worker_threads(1)
-            .build()
-            .unwrap();
-
-        let allow_origin = if disable_cors {
-            "null"
+    let mut rpc = RpcModule::new(());
+    rpc.merge(EthNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(EthTestNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(AnvilNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(EvmNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(DebugNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(NetNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(ConfigNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(ZksNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(Web3Namespace.into_rpc()).unwrap();
+
+    let cors_layers = tower::util::option_layer(if !disable_cors {
+        // `CorsLayer` adds CORS-specific headers to responses but does not do filtering by itself
+        let cors_layer = CorsLayer::new()
+            .allow_origin(AllowOrigin::exact(
+                cors_allow_origin.parse().expect("malformed allow origin"),
+            ))
+            .allow_headers([http::header::CONTENT_TYPE])
+            .allow_methods([Method::GET, Method::POST])
+            .expose_headers(Any);
+        // `HostFilterLayer` filters requests based on "Host" header
+        let host_filter_layer = if cors_allow_origin == "*" {
+            HostFilterLayer::disable()
         } else {
-            &cors_allow_origin
+            HostFilterLayer::new([cors_allow_origin]).expect("malformed allow origin")
         };
-        let mut builder = jsonrpc_http_server::ServerBuilder::new(io_handler)
-            .threads(1)
-            .event_loop_executor(runtime.handle().clone())
-            .cors(DomainsValidation::AllowOnly(vec![allow_origin.into()]));
+        Some((cors_layer, host_filter_layer))
+    } else {
+        None
+    });
+    let health_api_layer = tower::util::option_layer(if enable_health_api {
+        Some(ProxyGetRequestLayer::new("/health", "web3_clientVersion").unwrap())
+    } else {
+        None
+    });
+    let server_builder = ServerBuilder::default()
+        .http_only()
+        .set_http_middleware(
+            tower::ServiceBuilder::new()
+                .layer(cors_layers)
+                .layer(health_api_layer),
+        )
+        .set_rpc_middleware(RpcServiceBuilder::new().rpc_logger(100));
 
-        if enable_health_api {
-            builder = builder.health_api(("/health", "web3_clientVersion"));
-        }
+    let server = server_builder.build(addr).await.unwrap();
 
-        let server = builder.start_http(&addr).unwrap();
+    tokio::spawn(async move {
```

Comment:
A few things here:
1. Looks like sender/receiver was used as an "adapter" to create a tokio task (async) that would end when the task in a different _thread_ stops (sync), so we probably can remove it altogether now (since `server_handle` is already async).
2. Creating orphan tasks is generally not a good pattern.
3. If we are to move to a builder pattern, probably it makes sense to introduce the `run` method and make `tokio::spawn` the responsibility of the caller, so that there is a better split of responsibility.

Reply:
Couldn't do exactly what you described as `jsonrpsee` is weirdly designed in a way where they want to be the ones calling `tokio::spawn`. But I think the approach is close enough and addresses your concerns. PTAL